File: go/src/client/options.go
package client

import (
	"net/http"

	"github.com/radiustechsystems/sdk/go/src/transport"
)

// Option is a functional option for configuring a new Radius Client.
// It follows the functional options pattern for clean and extensible API configuration.
type Option func(*Options)

// Options contains configuration options for a new Radius Client.
// These options control how the client connects to and interacts with the Radius node.
type Options struct {
	// httpClient is the HTTP client used for making API requests
	httpClient *http.Client

	// interceptor is a function for modifying or monitoring JSON-RPC responses
	interceptor transport.Interceptor

	// logger is a function for debugging request/response cycles
	logger transport.Logf
}

// WithHTTPClient creates an option to set a custom HTTP client for the Radius Client.
// By default, the standard http.Client is used for HTTP requests.
//
// @param client Custom HTTP client implementing the http.Client interface
// @return An Option function that can be passed to New()
func WithHTTPClient(client *http.Client) Option {
	return func(o *Options) {
		o.httpClient = client
	}
}

// WithInterceptor creates an option to set a response interceptor for the Radius Client.
// This can be used to log, modify, or analyze responses from the Radius server.
// It's useful for debugging, testing, and to temporarily patch any issues in JSON-RPC responses.
//
// @param interceptor Function that can intercept and potentially modify JSON-RPC responses
// @return An Option function that can be passed to New()
func WithInterceptor(interceptor transport.Interceptor) Option {
	return func(o *Options) {
		o.interceptor = interceptor
	}
}

// WithLogger creates an option to set a logger for the Radius Client.
// This can be used to log JSON-RPC requests and responses for debugging or audit purposes.
// The logger receives the raw request and response bodies for inspection.
//
// @param logger Function that logs messages with format strings and variable arguments
// @return An Option function that can be passed to New()
func WithLogger(logger transport.Logf) Option {
	return func(o *Options) {
		o.logger = logger
	}
}

File: typescript/src/client/options.ts
import { HttpClient } from '../common';
import { Interceptor, Logf } from '../transport';

/**
 * A function that configures the Radius Client with a given set of options
 * Used for functional-style configuration of the client
 *
 * @param options The options object to modify with configuration settings
 */
export type ClientOption = (options: ClientOptions) => void;

/**
 * Configuration options for the Radius Client
 */
export interface ClientOptions {
  /**
   * Custom HTTP client implementation for making API requests
   * If not provided, the global fetch function will be used
   */
  httpClient?: HttpClient;

  /**
   * Response interceptor for modifying or monitoring JSON-RPC responses
   * Can be used to implement custom error handling or response transformation
   */
  interceptor?: Interceptor;

  /**
   * Logger function for debugging request/response cycles
   * Useful for troubleshooting connection or parsing issues
   */
  logger?: Logf;
}

/**
 * Creates an option to set a custom HTTP client for the Radius Client
 * By default, the global fetch function is used for HTTP requests
 *
 * @param httpClient Custom HTTP client implementing the HttpClient interface
 * @returns A ClientOption function that can be passed to Client.New()
 */
export function withHttpClient(httpClient: HttpClient): ClientOption {
  return (options: ClientOptions) => {
    options.httpClient = httpClient;
  };
}

/**
 * Creates an option to set a response interceptor for the Radius Client.
 * This can be used to log, modify, or analyze responses from the Radius server.
 * It's useful for debugging, testing, and to temporarily patch any issues in JSON-RPC responses.
 *
 * @param interceptor Function that can intercept and potentially modify JSON-RPC responses
 * @returns A ClientOption function that can be passed to Client.New()
 */
export function withInterceptor(interceptor: Interceptor): ClientOption {
  return (options: ClientOptions) => {
    options.interceptor = interceptor;
  };
}

/**
 * Creates an option to set a logger for the Radius Client.
 * This can be used to log JSON-RPC requests and responses for debugging or audit purposes.
 * The logger receives the raw request and response bodies for inspection.
 *
 * @param logger Function that logs messages with format strings and variable arguments
 * @returns A ClientOption function that can be passed to Client.New()
 */
export function withLogger(logger: Logf): ClientOption {
  return (options: ClientOptions) => {
    options.logger = logger;
  };
}
